processor/src/operations/io_ops.rs

```diff
@@ -41,7 +41,7 @@ where
         // get the address from the stack and read the word from current memory context
         let ctx = self.system.ctx();
         let addr = self.stack.get(0);
-        let word = self.chiplets.read_mem(ctx, addr);
+        let word = self.chiplets.read_mem(ctx, addr)?;
```

Comment:
I would probably perform the check for valid memory address here and then everywhere downstream we'd know that it is a 32-bit value.